torch/csrc/autograd/record_function.h

```diff
@@ -26,31 +28,79 @@ struct TORCH_API StringView {
   const char* str_ptr_;
 };
 
-using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
+enum class RecordFunctionState {
+  NO_CALLBACKS, // no callbacks registered, no need to use RecordFunction
+  NAME_ONLY, // callbacks need only basic information (function name)
+  NAME_AND_INPUTS, // callbacks need both name and function inputs
+};
+TORCH_API RecordFunctionState recordFunctionState();
 
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn, GetPackedInputsCallback cb = nullptr);
+  // Default constructor is used with before function called afterwards
+  RecordFunction() : state_(recordFunctionState()) {}
+
+  // Constructors with parameters initialize RecordFunction
+  // members and call start callbacks
+  explicit RecordFunction(Function* fn);
 
   explicit RecordFunction(
       std::string name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+      int64_t current_sequence_nr = -1);
 
   explicit RecordFunction(
       const char* name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+      int64_t current_sequence_nr = -1);
+
+  // before function initializes RecordFunction members and calls
+  // start callbacks;
+  // before function can be used only when RecordFunction is
+  // constructed with default constructor RecordFunction::RecordFunction()
+  void before(const char* name, int64_t sequence_nr = -1);
+  void before(std::string name, int64_t sequence_nr = -1);
+  void before(Function* fn, int64_t sequence_nr = -1);
+
+  template<typename F, typename... Args>
+  void before(
+      F fn,
+      int64_t current_sequence_nr,
+      const Args&... args) {
+    if (state_ == RecordFunctionState::NAME_AND_INPUTS) {
+      inputs_ = std::vector<c10::IValue>({args...});
+    }
+    before(fn, current_sequence_nr);
+  }
 
-  explicit RecordFunction(
-      std::string name,
-      GetPackedInputsCallback cb) : RecordFunction(name, -1, cb) {}
+  template<typename F, typename... Args>
+  void before(
+      F fn,
+      const c10::IValue& value,
+      const Args&... args) {
+    before(fn, -1, value, args...);
+  }
 
-  explicit RecordFunction(
-      const char* name,
-      GetPackedInputsCallback cb) : RecordFunction(name, -1, cb) {}
+  template<typename F>
+  void before(
+      F fn,
+      const at::ArrayRef<c10::IValue>& values) {
```

Comment:
I can make it optional but it seems ArrayRef is already lightweight

Reply:
`ArrayRef`s are supposed to be passed by value, never by reference